pmd-core/src/main/java/net/sourceforge/pmd/PMD.java

```diff
@@ -297,19 +319,58 @@ public static RuleContext newRuleContext(String sourceCodeFilename, File sourceC
      *            RuleContext
      * @param renderers
      *            List of {@link Renderer}s
+     *
+     * @deprecated Use {@link #processFiles(PMDConfiguration, List, Collection, Report, List)}
+     * so as not to depend on {@link RuleSetFactory}. Note that this sorts the list of data sources in-place,
+     * which won't be fixed
      */
+    @Deprecated
     public static void processFiles(final PMDConfiguration configuration, final RuleSetFactory ruleSetFactory,
-            final List<DataSource> files, final RuleContext ctx, final List<Renderer> renderers) {
+                                    final List<DataSource> files, final RuleContext ctx, final List<Renderer> renderers) {
+        // Note that this duplicates the other routine, because the old behavior was
+        // that we parsed rulesets (a second time) inside the processor execution.
+        // To not mess up error handling, we keep this behavior.
+
         encourageToUseIncrementalAnalysis(configuration);
         sortFiles(configuration, files);
         // Make sure the cache is listening for analysis results
         ctx.getReport().addListener(configuration.getAnalysisCache());
 
-        final RuleSetFactory silentFactory = new RuleSetFactory(ruleSetFactory, false);
+        final RuleSetFactory silentFactory = ruleSetFactory.toParser().warnDeprecated(false).toFactory();
         newFileProcessor(configuration).processFiles(silentFactory, files, ctx, renderers);
         configuration.getAnalysisCache().persist();
     }
 
+    /**
+     * Run PMD using the given configuration. This replaces the other overload.
+     *
+     * @param configuration Configuration for the run. Note that the files,
+     *                      and rulesets, are ignored, as they are supplied
+     *                      as parameters
+     * @param rulesets      Parsed rulesets
+     * @param files         Files to process
+     * @param report        Report in which violations are accumulated
+     * @param renderers     Renderers that render the report
+     *
+     * @throws RuntimeException If processing fails
+     */
+    public static void processFiles(final PMDConfiguration configuration,
```

Comment:
Should we mark this as experimental? Or would we just deprecate this as well, if we have a better PMD programmatic API?

Reply:
It may be possible to preserve compatibility with this signature in PMD 7. But also I think we should ideally keep a published entry point, that remains non-deprecated in the 6.x release cycle